docs/source/format/Columnar.rst

```diff
@@ -0,0 +1,1266 @@
+  between physical value arrays in a nested type structure. For
+  example, a ``List<T>``-type parent array has a T-type array as its
+  child (see more on lists below).
+* **Leaf node** or **leaf**: A primitive value array that may or may
+  not be a child array of some array with a nested type.
+
+Physical Memory Layout
+======================
+
+Arrays are defined by a few pieces of metadata and data:
+
+* A logical data type
+* A sequence of buffers
+* A length as a 64-bit signed integer. Implementations are permitted
+  to be limited to 32-bit lengths, see more on this below
+* A null count as a 64-bit signed integer
+* An optional **dictionary**, for dictionary-encoded arrays
+
+Nested arrays additionally have a sequence of one or more sets of
+these items, called the **child arrays**.
+
+Each logical data type has a well-defined physical layout. Here are
+the different physical layouts defined by Arrow:
+
+* **Primitive (fixed-size)**: a sequence of values each having the
+  same byte or bit width
+* **Variable-size Binary**: a sequence of values each having a variable
+  byte length. Two variants of this layout are supported using 32-bit
+  and 64-bit length encoding.
+* **Fixed-size List**: a nested layout each each value has the same
+  number of elements taken from a child data type.
+* **Variable-size List**: a nested layout type where each value is a
+  variable-length sequence of values taken from a child data type. Two
+  variants of this layout are supported using 32-bit and 64-bit length
+  encoding.
+* **Struct**: a nested layout type consisting of a collection of child
+  **fields** each having the same length
+* **Sparse** and **Dense Union**: a nested layout type representing a
+  sequence of values, each of which can have type chosen from a
+  collection of child array types.
+* **Null**: a sequence of all null values, having null logical type
+
+Buffer Alignment and Padding
+----------------------------
+
+Implementations are recommended allocate memory on aligned (8- or
+64-byte boundaries) and pad (overallocate) to a length that is a
+multiple of 8 or 64 bytes. When serializing Arrow data for
+interprocess communication, these alignment and padding requirements
+are enforced. If possible, we suggest that you prefer using 64-byte
+alignment and padding. Unless otherwise noted, padded bytes do not
+need to have a specific value.
+
+The alignment requirement follows best practices for optimized memory
+access:
+
+* Elements in numeric arrays will be guaranteed to be retrieved via aligned access.
+* On some architectures alignment can help limit partially used cache lines.
+
+The recommendation for 64 byte alignment comes from the `Intel
+performance guide`_ that recommends alignment of memory to match SIMD
+register width.  The specific padding length was chosen because it
+matches the largest known SIMD instruction registers available as of
+April 2016 (Intel AVX-512).
```

Comment:
ARM may not agree (https://www.linleygroup.com/mpr/article.php?id=11753). Though I don't think they have implemented the maximum width in hardware yet :-)